flink-runtime/src/test/java/org/apache/flink/runtime/dispatcher/cleanup/DispatcherResourceCleanerFactoryTest.java

```diff
@@ -0,0 +1,420 @@
+    }
+
+    private void assertJobManagerMetricGroupNotCleaned() {
+        assertThat(jobManagerMetricGroup.numRegisteredJobMetricGroups()).isEqualTo(1);
+    }
+
+    private void assertLocalCleanupTriggered() {
+        assertThat(jobManagerRunnerRegistry.getLocalCleanupFuture()).isCompleted();
+        assertThat(jobGraphWriter.getLocalCleanupFuture()).isCompleted();
+        assertThat(blobServer.getLocalCleanupFuture()).isCompleted();
+        assertThat(highAvailabilityServices.getLocalCleanupFuture()).isCompleted();
+    }
+
+    private void assertGlobalCleanupTriggered() {
+        assertThat(jobManagerRunnerRegistry.getGlobalCleanupFuture()).isCompleted();
+        assertThat(jobGraphWriter.getGlobalCleanupFuture()).isCompleted();
+        assertThat(blobServer.getGlobalCleanupFuture()).isCompleted();
+        assertThat(highAvailabilityServices.getGlobalCleanupFuture()).isCompleted();
+    }
+
+    private void assertJobManagerMetricGroupCleaned() {
+        assertThat(jobManagerMetricGroup.numRegisteredJobMetricGroups()).isEqualTo(0);
+    }
+
+    private static class AbstractTestingCleanableResource
+            implements LocallyCleanableResource, GloballyCleanableResource {
+
+        private final CompletableFuture<JobID> localCleanupFuture = new CompletableFuture<>();
+        private final CompletableFuture<JobID> globalCleanupFuture = new CompletableFuture<>();
+
+        @Override
+        public void globalCleanup(JobID jobId) {
+            throw new UnsupportedOperationException("Synchronous globalCleanup is not supported.");
+        }
+
+        @Override
+        public void localCleanup(JobID jobId) {
+            throw new UnsupportedOperationException("Synchronous localCleanup is not supported.");
+        }
+
+        @Override
+        public CompletableFuture<Void> localCleanupAsync(JobID jobId, Executor ignoredExecutor) {
+            localCleanupFuture.complete(jobId);
+
+            return FutureUtils.completedVoidFuture();
+        }
+
+        @Override
+        public CompletableFuture<Void> globalCleanupAsync(JobID jobId, Executor ignoredExecutor) {
+            globalCleanupFuture.complete(jobId);
+
+            return FutureUtils.completedVoidFuture();
+        }
+
+        public CompletableFuture<JobID> getLocalCleanupFuture() {
+            return localCleanupFuture;
+        }
+
+        public CompletableFuture<JobID> getGlobalCleanupFuture() {
+            return globalCleanupFuture;
+        }
+    }
+
+    private static class CleanableJobGraphWriter extends AbstractTestingCleanableResource
+            implements JobGraphWriter {
+
+        @Override
+        public void putJobGraph(JobGraph jobGraph) {
+            throw new UnsupportedOperationException("putJobGraph operation not supported.");
+        }
+    }
+
+    private static class CleanableHighAvailabilityServices extends AbstractTestingCleanableResource
```

Comment:
Why can't we reuse `TestingHighAvailabilityServices` here? Same applies to other classes in this test.

The reasoning behind this is very same as with using `mockito`, if you decide to change something in the interface you have to go trough all implementations one by one.

Reply:
fair point. I refactored it.